server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -573,12 +578,43 @@ public void addAllocFailuresResetListenerTo(ClusterService clusterService) {
         });
 
         clusterService.addListener((changeEvent) -> {
-            if (changeEvent.nodesAdded() && changeEvent.state().getRoutingNodes().hasAllocationFailures()) {
+            if (shouldResetAllocationFailures(changeEvent)) {
                 taskQueue.submitTask("reset-allocation-failures", (e) -> { assert MasterService.isPublishFailureException(e); }, null);
             }
         });
     }
 
+    /**
+     *  We should reset allocation/relocation failure count to allow further retries when:
+     *
+     *  1. A new node joins the cluster.
+     *  2. A node shutdown metadata is added that could lead to a node being removed or replaced in the cluster.
+     *
+     * Note that removing a non-RESTART shutdown metadata from a node that is still in the cluster is treated similarly and
+     * will cause resetting the allocation/relocation failures.
+     */
+    private boolean shouldResetAllocationFailures(ClusterChangedEvent changeEvent) {
+        final var clusterState = changeEvent.state();
+        boolean hasAllocationFailures = clusterState.getRoutingNodes().hasAllocationFailures();
+        boolean hasRelocationFailures = clusterState.getRoutingNodes().hasRelocationFailures();
```

Comment:
It might just be me not familiar with the code. But does the `reset-allocation-failures` task clear the relocation failures? Or does it not matter?

Reply:
[Clears both](https://github.com/elastic/elasticsearch/blob/bedeb0414a7bdcfdca82cf78471c561bc979e7b0/server/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java#L1301). As for the naming see [this comment](https://github.com/elastic/elasticsearch/pull/119968/files#r1916722015).

Reply:
Ah ok thanks.